Create how blockchain works page in  Lisk docs - Closes #1467 (#1475)
LiskHQ/lisk-docs: docs/modules/ROOT/pages/intro/how-blockchain-works.adoc

= How blockchain works
Chris Braithwaite <christopher.braithwaite@lightcurve.io> Mona Bärenfänger <mona@lightcurve.io>

:description: The How blockchain works page describes in more detail the functionalities of a blockchain.

:toc: preamble
:idprefix:
:idseparator: -
:imagesdir: ../../assets/images
//:experimental::toc:
:page-previous: /root/intro/what-is-blockchain.html
:page-previous-title: What is blockchain
:page-next: /root/intro/lisk-products.html
:page-next-title: Lisk products

:url_p2p_architecture: understand-blockchain/lisk-protocol/network.adoc
:url_state_store: understand-blockchain/modules-assets.adoc#the-state-store

// :url_hello: build-blockchain/create-blockchain-app.adoc

=== Overview

> As described on the previous page from a high-level perspective blockchain is basically a 'digital ledger' which has many functionalities and use cases.

Author: Done 👍

* Consensus Driven – A consensus model independently verifies each block on the blockchain, and provides certain rules for validating a block, which may utilise key resources such as computing power, in order to demonstrate that this has been accomplished.
For example, with regard to Bitcoin this is known as the mining process.
There are different types of consensus models which will be covered further in this section, however suffice to say that a consensus model is able to function independently, and does not require any form of central authority.

* Immutability – A blockchain can be considered as a permanent, unchangeable, and irreversible record of transactions.
Hence, once a block is created and added to the blockchain it ensures the transaction is completely trustable.

* Transparency – A blockchain by its very nature provides total open transparency.
A full transaction history  can be viewed by any party, as it is in effect an open file.
Therefore another highly beneficial advantage is that a blockchain creates provenance, as all transactions can be easily audited from the first to the last entry.

> === Consensus mechanisms

Author: Good point...Now done 👍

It is important to understand consensus within the blockchain ecosystem.
Basically, a consensus mechanism is a methodology deployed by nodes to agree on, and validate the data content of a block before it becomes part of a blockchain.
The block containing this data is then distributed to all the nodes within the network.
The consensus algorithm is responsible for ensuring that only 1 copy of the data encapsulated in each block exists within each node.

Various consensus mechanisms exist today, however the 4 main mechanisms that are widely used consist of the Proof-of-Work, Proof-of-Stake, Delegated-Proof-of-Stake, and Byzantine-Fault-Tolerance mechanisms, which are explained below:

> === Consensus algorithms

Author: OK, I have added some relevant links... 👍

To put it in a nutshell, a consensus algorithm can be thought of as a necessary procedure which results in ensuring all peers of a blockchain network come to a common agreement of the current state of the network.

=== Proof-of-Work (PoW)

Proof of Work is probably the most well-known and established consensus mechanism, as it is utilized by Bitcoin.

> However, when compared to the other consensus mechanisms, it is somewhat inefficient as it requires a high amount of energy and processing power to reach a consensus, therefore being rather cumbersome and expensive to operate.

Author: I have updated this 👍

> Basically to validate transactions it requires the nodes to solve an arbitrary mathematical problem or complex equations, which prevents the network being hacked, and also provides a high level of security.

Author: Done 👍

It is also widely used for mining cryptocurrencies and minting new tokens.

=== Proof-of-Stake (PoS)
To explain PoS briefly, it allows the users to stake an asset/token which in turn opens up the possibility to be chosen as a validator of a new block, which then allows transaction fees, or newly minted tokens to be collected from the block as a reward.
Staking is defined as an amount of tokens/assets that are held by the node.

> The proof-of-stake-mechanism uses a random algorithm for consensus building and eliminates the need for complex mathematical calculations, hence reducing the overall computing power and energy required.

Author: It should now be clearer 👍

This consensus algorithm used in DPoS utilizes a unique election system that is able to select nodes which are able to perform block verification.
In this system, generally  delegates are voted in based upon their reputation.
Each user who holds a minimum of one token/coin with the DPoS blockchain has the ability to vote specifically for the nodes that they want to perform the transaction validations.
This is the consensus mechanism used by Lisk, and hence every Lisk token holder is able to vote for delegates, and dependent on these votes, certain delegates are allowed to add blocks to the blockchain in a specific order.
The delegates are actually normal Lisk accounts that have performed a delegate registration transaction.
The whole process can be considered as the most democratic method of voting, as it does not rest on the users that possess the highest amount of tokens, as just because a particular user has many tokens, that does not authorize this user to validate and confirm transactions.
All the users that maintain tokens in their accounts are able to select a group of delegates to perform this task.

> Furthermore, the DPoS has the advantage of being able to execute transactions and verifications much faster, which results in higher scalability.

Author: Done 👍

The key requirements for BFT must be accomplished by the nodes within a blockchain network, therefore, it is imperative they are deterministic and must begin with the same state for practical BFT.

* Liveness: New blocks can still be finalized on the network, even in the case whereby one third of the active delegates are offline.

The process of being able  to tolerate an amount of  Byzantine process failures or crashes, whilst ensuring the network continues to function is defined as Byzantine fault tolerant (BFT).
To achieve BFT consensus the following requirements must be met:

1. Termination: An output is decided by every non-faulty process.
2. Agreement: The same output is decided by every non-faulty process.
3. Validity: Every process starts with the same input.
4. Integrity: The consensus value and all non-faulty process decisions achieved in point 2 above, need to have been put forward by some non-faulty process.

=== Peer-2-Peer networks

The usage of a peer-to-peer model in blockchain offers various benefits as compared to a standard client-server based system.

> By its very nature a P2P system offers a high level of security, as it is virtually immune to any Denial-of-Service (DoS, attacks), any also any censorship by a central governing body (for example, assets cannot be frozen or manipulated by any entity).

Author: I have reworded it 👍

> By its very nature a P2P system offers a high level of security, as it is virtually immune to any Denial-of-Service (DoS, attacks), any also any censorship by a central governing body (for example, assets cannot be frozen or manipulated by any entity).

Author: Now updated 👍

To summarise, if blockchain technology is deployed using a P2P network architecture, then this will result in greater freedom, security, immutability, and enhanced decentralization.

==== Unstructured P2P networks

In an unstructured P2P network there is no organization, hence the nodes are able to communicate and connect randomly.
Such networks are more suited towards social platforms, or any type of system which experiences high churn rates.
However, such networks require a high amount of processing power and may incur long delays.

> ==== Structured P2P networks

Reviewer: I don't think it is relevant to explain structured and hybrid P2P networks in much detail, unless they are used in other popular blockchain networks. At Lisk we have an unstructured partial mesh network, so it would be nice explain that to the reader.

Also, diagrams can be really helpful to explain differences between the different networks without the need of too many words.

Author: Good point. I have added an extra paragraph to explain unstructured partial mesh networks and a link to our P2P architecture 👍

Lisk uses what is known as an unstructured P2P network, whereby the nodes randomly connect to each other.
This methodology is highly beneficial as the nodes then broadcast their new transactions, which results in synchronizing all their local copies of the blockchain.
The xref:{p2p_architecture}[Lisk P2P architecture] is covered in more detail in the Lisk Protocol section.

=== Cryptography

Cryptography is not a new concept, and ultimately is used to ensure secure communication between 2 parties can be established over an unsecure connection.
This can be accomplished by using an encryption mechanism as depicted in the example diagram below, whereby a plain text document is sent as a ciphered version to the recipient.
The recipient can only decrypt the ciphered text if he or she holds the identical symmetrical encryption key.
Therefore, this can be transmitted over any unsecure medium, as regardless of any third party or malicious actors intercepting this, they would not be able to decrypt the text, rendering it useless to them.

> image:intro/cryptography.png[]

Author: Good point! I have revamped the Cryptography section, and also built an additional diagram for asymmetric encryption as well 👍

* Hash Functions:
This function does not utilise any keys, as it takes the contents of the plain text and deploys a cipher, which is used to generate a hash value of a fixed length from the plain text.
Hence, it is virtually impossible for the contents of this plain text to be unravelled from the cipher text.
Therefore, hashing does not only provide the required security, as described earlier on the previous page, it is also deterministic and has the ability to provide immutability as well, which as we have learnt is highly beneficial.

Lisk maintains a cryptography package which contains all the cryptographic functionalities required when interacting with the Lisk ecosystem, and can be used on both the server and client side.

=== State machine

A State machine is considered to be concept whereby the definition relates to a machine that can have multiple states, however only one state is possible at any one given time.
Hence, a state in this case refers to the current state of the blockchain system and its transactions that are responsible for triggering state transitions.
With regard to a blockchain system, it can be deemed as a deterministic, replicated state machine.

> image::intro/state-machine.png[]

Reviewer: I am not 100% sure but I think blocks alone can also alter the state of the blockchain?

Author: @Incede Rishi, Can you clarify this point please, so if necessary I can update the diagram accordingly? Many thanks.

Author: Good point, I have now updated this 👍

Reviewer: Just for clarification, I think this introduction is already for v6, or am I missing something?

Reviewer 2: I noticed the terms `account state` `chain state` & `consensus state` being used which are only valid for v5, hence my assumption. If it's intended for v6 please remove these terms & related content or update them with v6 parameters. 

In any case, the fact that a new block always changes the state is still valid.

Author: I have left them as they are for the moment, as if I am correct they are not all agreed and accepted as of yet? 
I saw there was a Trello card on V6 terminologies with a meeting scheduled for 22 July.

Reviewer: I agree with @Incede , these terms are only valid for v5, and also they are specific for the Lisk SDK, so I would remove them here.

The state transition refers to the changes that occur in the state machine after a specific event has occurred.

> The state store can mutate the state of the actual blockchain data, and it can also retrieve data from the blockchain.

Reviewer: The state store is a special data structure we use in the Lisk SDK, it is not exactly the same as a state machine, so I wouldn't explain the state store here at this place

Author: Agreed...OK, I have now removed this sentence 👍

* Consensus Driven – A consensus model independently verifies each block on the blockchain, and provides certain rules for validating a block, which may utilize key resources such as computing power, in order to demonstrate that this has been accomplished.
For example, with regard to Bitcoin, this is known as the mining process.
There are different types of consensus models which will be covered further in this section, however, suffice to say that a consensus model is able to function independently, and does not require any form of central authority.

* Immutability – A blockchain can be considered a permanent, unchangeable, and irreversible record of transactions.
Hence, once a block is created and added to the blockchain it ensures the transaction is completely trustable.

* Transparency – A blockchain by its very nature provides total open transparency.
A full transaction history can be viewed by any party, as it is in effect an open file.
Therefore another highly beneficial advantage is that a blockchain creates provenance, as all transactions can be easily audited from the first to the last entry.

> == Consensus mechanisms

Reviewer: Please note down the following points as checklists for each consensus algorithm, to summarize:

- What are the advantages of this consensus?
- What are the possible disadvantages?

Author: Done 👍

> == Consensus mechanisms

Author: As discussed, I have now created a Trello card in the Design request board 👍

It is important to understand consensus within the blockchain ecosystem.
Basically, a consensus mechanism is a methodology deployed by nodes to agree on, and validate the data content of a block before it becomes part of a blockchain.
The block containing this data is then distributed to all the nodes within the network.
The consensus algorithm is responsible for ensuring that only 1 copy of the data encapsulated in each block exists within each node.
To put it in a nutshell, a consensus mechanism can be thought of as a necessary procedure that results in ensuring all peers of a blockchain network come to a common agreement on the current state of the network.

Various consensus mechanisms exist today, however, the 4 main mechanisms that are widely used consist of the *Proof-of-Work*, *Proof-of-Stake*, *Delegated-Proof-of-Stake*, and *Byzantine-Fault-Tolerance* mechanisms, which are explained below:

> === Proof-of-Work (PoW)

Author: Done 👍

To summarise,  a xref:{url_p2p_networks}[P2P network architecture] offers numerous benefits as compared to a traditional client-server-based network and results in greater freedom, improved security, and enhanced decentralization.

=== Unstructured P2P networks

In an unstructured P2P network there is no organization, hence the nodes are able to communicate and connect randomly.
Such networks are more suited to social platforms, or any type of system which experiences high churn rates.
However, such networks require a high amount of processing power and may incur long delays.

=== Structured P2P networks

A structured network can be considered the opposite of an unstructured network and is organized in a manner whereby the nodes are able to utilize efficient search mechanisms.
This can be performed by the nodes using a hash function.
However, although they are considered to be more efficient, they can be considered somewhat centralized.

=== Hybrid P2P networks

The hybrid P2P networks are actually a combination of the client-server model and the peer-to-peer architecture.

> This can offer the best of both worlds, and operate more efficiently as they generally contain an index/central server which can also connect and provide connections between the network nodes.

Reviewer: I don't think this is relevant, can be removed

Author: Done 👍

> === Unstructured partial mesh networks

Reviewer: This should be moved under `Unstructured P2P networks` section, because it is a subcategory of that network type

> === Unstructured partial mesh networks

Reviewer: Please note that we have already a section in our existing docs talking about the p2p network: https://liskhq.github.io/lisk-docs/understand-blockchain/index.html#p2p-unstructured-partial-mesh-network

Try to remove redundant content and update this, or the existing content in a way that is complements each other in a nice way.

Author: Done 👍

Lisk uses what is known as an unstructured P2P network, whereby the nodes randomly connect to each other, and are all considered equal participants in the network.
This methodology is highly advantageous as with a partial mesh topology not all nodes are connected to each other, and are only connected to a subset of nodes present in the network.
This increases the robustness and scalability, coupled with increased security.
Lisk utilizes the unstructured partial network methodology, and the following link in the *Lisk Elements* section shows the xref:{p2p_architecture}[Lisk P2P architecture] in more detail.

> == Cryptography

Author: Done 👍

This is the simplest method, as 1 common key is used for both the encryption and decryption process.
In this case, it is necessary to ensure the transfer of the common key can be performed safely from the sender to the recipient.
It is also referred to as secret-key cryptography as shown in the illustration below.

image:intro/symmetric-encryption.png[]

=== Asymmetric Key
This type of encryption functions by using a pair of keys.
This comprises an encryption key, and a decryption key, and is more commonly known as a public key and a private (or secret), key.
Basically, the algorithm deployed for this method generates both a secret, key and a unique public key.
The secret key as its name implies, is kept secret, and the public key is openly shared.
Furthermore, the asymmetric encryption method has an additional element of security, although the symmetric method of encryption is faster, nevertheless, they are both very effective.
However, it is important to note that there are two methods of cryptography algorithms that are deployed within the blockchain, which are the asymmetric model described here, and the hashing function described in the following paragraph below.

> As discussed, the encryption techniques play an important role here, as it is critical for a blockchain to use the most secure methodology, which ensures the key functions can be securely managed, such as transaction authentication, digital signatures, and identity management.

Reviewer: I think it would be great to structure this information better for the reader, not having only big chunks of text to digest. The main information the reader want to get here is

- what are the different encryption methods
- where are they used in blockchain context, followed by a simple example

It is quite common with regard to the management of cryptocurrencies that with the asymmetric model of encryption, the public key is generally the actual address that 'contains' the tokens or coins, and is publicly viewable.
Subsequently, the secret key is used by the holder of the tokens or coins to access the address, and hence be able to authorize and then perform any actions required.

> image:intro/asymmetric-encryption.png[]

Author: I have had a look at numerous depictions of how to explain this from a graphical image perspective, and imho I prefer the 2 key image. In addition, it maintains the uniformity with the first 1 key image describing the symmetric encryption, so I would prefer to leave it as it is. :pray:

:url_pos: https://www.investopedia.com/terms/p/proof-stake-pos.asp
:url_p2p_networks: https://www.blockchain-council.org/blockchain/blockchain-role-of-p2p-network/
:url_hashing: https://www.onlinehashcrack.com/how-to-hashing-in-blockchain-explained.php
:url_account-store: lisk-sdk::references/lisk-elements/chain.adoc#state-store-and-database-mechanism
:url_voting-mechanism: https://blockchain-academy.hs-mittweida.de/courses/blockchain-introduction-technical-beginner-to-intermediate/lessons/lesson-20-introduction-and-basic-functionality-of-delegated-proof-of-stake/topic/voting-in-dpos/
:url_lisk-products: intro/lisk-products.adoc

== Overview

As described on the previous xref:{url_blockchain}[What is blockchain] page, from a high-level perspective a blockchain a 'digital ledger' which has many functionalities and use cases.
It can also be thought of as a digital methodology to store data, whereby the data is encapsulated in blocks, and once the blocks are chained together they cannot be changed or altered, thus making them immutable.
Once a block of data is added to the blockchain it will remain publicly viewable to anyone at any time.

> This is a highly beneficial system as it enables permanent records to be kept, enabling the storage of almost any type of record.

Author: Yes, this is true. However, I didn't really want to go into too much depth on this, although this is a valid point, so I reworded this sentence so that It reads " 'in general', it enables permanent records to be kept" 👍

> This is a highly beneficial system as it enables permanent records to be kept, enabling the storage of almost any type of record.

Author: I have also updated this further in a bit more detail.

For example, it could be virtually anything such as medical records, financial transactions, property rights, etc.

To start to break down how the blockchain works, and before we delve into the detailed contents of a block, firstly we need to understand how the blocks are linked together, this is accomplished by using cryptography.
Therefore, each block contains a cryptographic hash of the previous block, the transaction data,  and a timestamp.
This will be explained in more detail further in this section.
As previously mentioned once created, the blocks are immutable and cannot be changed retroactively.
The only way this could be achieved would be by altering the subsequential blocks, which would require the consensus of the entire network.

== Key characteristics of blockchain

There are 4 core characteristics attributed to the blockchain which are covered below:

> * Decentralization – The contents of each block in the blockchain can be accessed by any node on the network, as they are essentially network copies that are present on all participating nodes.

Author: OK, I have reworded this paragraph to state the following:
 ''Decentralization -  This can be defined as storing the blocks over a network of nodes, and not in any one central location. The contents of each block in a public blockchain can be accessed by any node on the network, as they are essentially network copies that are present on all participating nodes.
Therefore, there is no central entity that holds each block in the chain, so this results in decentralization."

> * Decentralization – The contents of each block in the blockchain can be accessed by any node on the network, as they are essentially network copies that are present on all participating nodes.

Author: A minor update  to the last sentence: 'Therefore, there is no central entity that holds any control to change each block in the chain, so this results in decentralization'.

> Therefore, there is no central entity that holds each block in the chain, so this creates decentralization.

Author: Yes, good point. I have adapted the last sentence to reflect this point "Therefore, there is no central entity that holds any control to change each block in the chain, so this results in decentralization".

* Consensus Driven – A consensus model independently verifies each block on the blockchain, and provides certain rules for validating a block, which may utilize key resources such as computing power, in order to demonstrate that this has been accomplished.
For example, with regard to Bitcoin, this is known as the mining process.
There are different types of consensus models which will be covered further in this section, however, suffice to say that a consensus model is able to function independently, and does not require any form of central authority.

> * Immutability – A blockchain can be considered a permanent, unchangeable, and irreversible record of transactions.

Author: Updated to Immutability – A blockchain can be considered an unchangeable, and irreversible record of transactions.

== Consensus mechanisms

It is important to understand consensus within the blockchain ecosystem.
Basically, a consensus mechanism is a methodology deployed by nodes to agree on, and validate the data content of a block before it becomes part of a blockchain.
The block containing this data is then distributed to all the nodes within the network.
The consensus algorithm is responsible for ensuring that only 1 copy of the data encapsulated in each block exists within each node.
To put it in a nutshell, a consensus mechanism can be thought of as a necessary procedure that results in ensuring all peers of a blockchain network come to a common agreement on the current state of the network.

Various consensus mechanisms exist today, however, the 4 main mechanisms that are widely used consist of the *Proof-of-Work*, *Proof-of-Stake*, *Delegated-Proof-of-Stake*, and *Byzantine-Fault-Tolerance* mechanisms, which are explained below:

=== Proof-of-Work (PoW)

* What are the advantages of PoW?

The major advantage regarding Proof-of-Work is that it is highly scalable, and generally very reliable.

> It is also effective for crypto mining, minting new tokens, and validating transactions.

Author: Updated to ''It is also effective for minting new tokens, and validating transactions''.

In essence, it allows consensus to be reached regardless if some of the nodes fail.

Firstly, there are 3 key features whereby BFT is able to improve the blockchain, and they are listed below:

* Safety: If 2 conflicting blocks occur on the network, then assuming two-thirds of the active delegates adhere honestly to the protocol, then these 2 conflicting blocks will not be finalized on the blockchain.

* Accountability: In the case whereby the protocol is violated by the delegate, they will be held responsible for this.
The key requirements for BFT must be accomplished by the nodes within a blockchain network, therefore, it is imperative they are deterministic and must begin with the same state for practical BFT.

* Liveness: New blocks can still be finalized on the network, even in the case whereby one third of the active delegates are offline.

The process of being able to tolerate a number of Byzantine process failures or crashes, whilst ensuring the network continues to function is defined as Byzantine-fault-tolerant.
To achieve BFT consensus the following requirements must be met:

> 1. Termination: An output is decided by every non-faulty process.

Author: Now updated 👍

2. Agreement: The same output is decided by every non-faulty process.
3. Validity: Every process starts with the same input.
4. Integrity: The consensus value and all non-faulty process decisions achieved in point 2 above, need to have been put forward by some non-faulty process.

== Peer-2-Peer networks

The usage of a peer-to-peer model in blockchain offers various benefits as compared to a standard client-server-based system.

> By its very nature, a P2P decentralized system offers a high level of security and does not require any third-party intermediaries, and in addition, assets cannot be frozen or manipulated by any entity.

Author: Reworded to 'By its very nature, in a P2P decentralized system, assets cannot be frozen or manipulated by a centralized entity'.

Therefore, when a transaction is created, this has to be digitally signed with the secret key.
Once this has been received by the blockchain network, it can be verified with the same public key, therefore this proves the authorization of the transaction is genuine by the owner of the secret key, or any entity with knowledge of the secret key.
Hence, transactions can be authenticated without having to reveal the identity of the account owner.
Subsequently, the public key is used for identity management and the account address, and the secret key is used for the passphrase of the user account to access the address, and hence be able to authorize and then perform any actions required.

image:intro/asymmetric-encryption.png[]

=== Hash Functions
This function does not utilize any keys, as it takes the contents of the plain text and deploys a cipher, which is used to generate a hash value of a fixed length from the plain text.
Hence, it is virtually impossible for the contents of this plain text to be unraveled from the cipher text.

> Therefore, {url_hashing}[hashing^] does not only provide the required security, as described earlier on the previous page, it is also deterministic and has the ability to provide immutability as well, which as we have learned is highly beneficial.

Author: Therefore, {url_hashing}[hashing^] hash function outputs can be used to efficiently to verify the inputs without revealing the input itself., as described earlier on the previous page, it is also deterministic and is the key component in providing immutability of the blockchain data as well, which as we have learned is highly beneficial.

Lisk maintains a cryptography package that contains all the cryptographic functionalities required when interacting with the Lisk ecosystem and can be used on both the server and client-side.

== State machine

A State machine is considered to be a concept whereby the definition relates to a machine that can have multiple states, however only one state is possible at any one given time.
Hence, a state in this case, refers to the current state of the blockchain system and its transactions that are responsible for triggering state transitions.
With regard to a blockchain system, it can be deemed as a deterministic, replicated state machine.

image::intro/state-machine.png[]

The state transition refers to the changes that occur in the state machine after a specific event has occurred.

> Therefore, it is best thought of as a temporary data structure that holds a temporary state while processing a block.

Author: Good point...now reworded.

> Furthermore, it maintains a temporary state that exists during the processing of a block, as it exposes an interface that enables and results in the finalization of the snapshots.

Author: Agreed, I have removed this part of the sentence.
Furthermore, it maintains a temporary state that exists during the processing of a block.

> From a high-level perspective, the state store can be broken down into 3 separate states, namely the Accounts, the Chain, and the Consensus.

Author: Agreed , so have removed line 242

:url_blockchain: intro/what-is-blockchain.adoc
:url_pow: https://www.investopedia.com/terms/p/proof-work.asp
:url_pos: https://www.investopedia.com/terms/p/proof-stake-pos.asp
:url_p2p_networks: https://www.blockchain-council.org/blockchain/blockchain-role-of-p2p-network/
:url_hashing: https://www.onlinehashcrack.com/how-to-hashing-in-blockchain-explained.php
:url_account-store: lisk-sdk::references/lisk-elements/chain.adoc#state-store-and-database-mechanism
:url_voting-mechanism: https://blockchain-academy.hs-mittweida.de/courses/blockchain-introduction-technical-beginner-to-intermediate/lessons/lesson-20-introduction-and-basic-functionality-of-delegated-proof-of-stake/topic/voting-in-dpos/
:url_lisk-products: intro/lisk-products.adoc

== Overview

As described on the previous xref:{url_blockchain}[What is blockchain] page, from a high-level perspective a blockchain is a 'digital ledger' which has many functionalities and use cases.
It can also be thought of as a digital methodology to store data, whereby the data is encapsulated in blocks, and once the blocks are chained together they cannot be changed or altered, thus making them immutable.

> Once a block of data is added to the blockchain it will remain publicly viewable to anyone at any time, assuming it is not a permissioned or private blockchain.

Author: Done. I also included permissioned blockchains and links 👍

== Consensus mechanisms

It is important to understand consensus within the blockchain ecosystem.
Basically, a consensus mechanism is a methodology deployed by nodes to agree on, and validate the data content of a block before it becomes part of a blockchain.
The block containing this data is then distributed to all the nodes within the network.
The consensus algorithm is responsible for ensuring that the exact same copy of the data encapsulated in each block exists within each node.
To put it in a nutshell, a consensus mechanism can be thought of as a necessary procedure that results in ensuring all peers of a blockchain network come to a common agreement on the current state of the network.

Various consensus mechanisms exist today, however, the 4 main mechanisms that are widely used consist of the *Proof-of-Work*, *Proof-of-Stake*, *Delegated-Proof-of-Stake*, and *Byzantine-Fault-Tolerance* mechanisms, which are explained below:

=== Proof-of-Work (PoW)

image::intro/pow.png[ align="center" ,600]

* What are the advantages of PoW?

The major advantage regarding Proof-of-Work is that it is highly scalable, and generally very reliable.

> It is also effective for minting new tokens and validating transactions.

Author: OK, I have added more content to cover this in more detail. There is a link for a more in-depth explanation of PoW at the end of this section 👍

* What are the disadvantages of PoW?

On the other hand, the disadvantages consist of PoW being very energy intensive, since high amounts of energy are consumed in the mining process.
In addition, it can be vulnerable to malicious attacks (e.g. the well known 51% attack).

> Proof of Work is probably the most well-known and established consensus mechanism, as it is utilized by Bitcoin.

Author: Ah yes...good point, and well spotted Mona! 👍

Hence, this results in the miners receiving a reward for this in the form of tokens or coins.
A further more in-depth explanation of PoW can be found xref:{url_pow}[here].

=== Proof-of-Stake (PoS)

image::intro/pos-v3.png[ align="center" ,600]

* What are the advantages of PoS?

Proof-of-Stake is far more efficient than PoW as it does not require any energy intensive computer hardware, and does not require such complex cryptographic mathematical problems to be solved to complete the mining process.
In addition, PoS is more energy efficient than PoW, which requires miners to use a high level of electricity to complete their tasks.

* What are the disadvantages of PoS?

One of the well known disadvantages relates to the fact that the mining power in PoS is determined by the number of tokens that a validator has staked, therefore it is often said that this tends to benefit the more wealthy participants.
Hence, users that stake more tokens have a higher chance of being chosen to forge new blocks.

> To explain PoS briefly, it allows the users to stake an asset/token which in turn opens up the possibility to be chosen as a validator of a new block, which then allows transaction fees, or newly minted tokens to be collected from the block as a reward.

Author: Done, and reworded accordingly 👍

Staking is defined as a number of tokens/assets that are held by the node.
The proof-of-stake-mechanism uses an algorithm designed to select users that have the highest stakes as validators, this results in motivating the highest stakeholders to ensure a transaction is executed, as logic dictates that the users with the highest amount of tokens or coins have the most to lose, therefore it is in their interest to ensure the network continues to grow.
This is highly beneficial for consensus building and eliminates the need for complex mathematical calculations, hence reducing the overall computing power and energy required.
Alternatively, the nodes that hold the most tokens/assets have the power to validate the transactions, which in turn results in this node receiving a reward, usually in the form of newly minted tokens.
Further more detailed information regarding PoS can be found {url_pos}[here^].

=== Delegated-Proof-of-Stake (DPoS)

//  Awaiting DPOS image -  image::intro/dpos1.jpeg[ align="center" ,600]

* What are the advantages of DPoS?

> The DPoS mechanism possesses an open and transparent voting system, and furthermore provides faster transaction processing times than PoW and PoS.

Author: Added more content and a link to substantiate this 👍

This in turn is beneficial for many applications that require a high level of scalability.
Moreover, should any malicious activity on the network be discovered, the participants can vote to have the offending delegate removed immediately, therefore providing a good incentive for delegates to behave correctly.

> Finally, it is energy efficient and requires less computing power and cumbersome hardware.

Author: This is covered in the new link I added in the preceding paragraph which gives the reader the opportunity to read about DPoS in much more depth.

* What are the disadvantages of DPoS?

DPos is a partially centralized system, therefore the delegates wth more tokens tend to have more power in the network.

> Furthermore, it can be vulnerable to attacks as often there are only minimal participants in charge of keeping the network functional.

Author: I have added a bit more content here 👍

DPoS works in a similar fashion to PoS, however, one of the key advantages is that it utilizes a delegation and voting mechanism, which in turn incentivizes the users to use their staked collateral to secure the network.
To allow this mechanism to be both efficient and effective at performing transaction validations, various different components of delegated proof of stake exist.

> Ultimately, the DPoS is considered the most advantageous mechanism here, as it was designed to negate most of the limitations that exist with PoW and PoS.

Reviewer: Can you provide proof for this statement?

> This consensus algorithm used in DPoS utilizes a unique election system that is able to select nodes that can perform block verification.

Reviewer: ```suggestion
DPoS utilizes a unique election system that is able to select users that can perform block verification.
```

Reviewer 2: Users(validators) perform block verification via nodes so probably the original fits better.

> In this system, generally delegates are voted in based on their reputation.

Reviewer: What are delegates in DPoS context? How can you become a delegate? Please explain, when introducing this new term to the reader. You can make important terms in the text also bold to highlight them to the reader. The first time you mention `delegate`, it should be bold, and it should be followed by an explanation f the term.

Author: Good point...I have added in more content to cover this 👍

> Each user who holds a minimum of one token/coin with the DPoS blockchain has the ability to vote specifically for the nodes that they want to perform the transaction validations.

Reviewer: ```suggestion
Each user who holds a minimum of one token/coin with the DPoS blockchain has the ability to vote specifically for the delegates that they want to perform the transaction validations.
```

> *Delegated-Proof-of-Stake* is the consensus mechanism used by Lisk, and hence every Lisk token holder is able to vote for delegates, and dependent on these votes, certain delegates are allowed to add blocks to the blockchain in a specific order.

Reviewer: ```suggestion
Dependent on these votes, certain delegates are allowed to add blocks to the blockchain in a specific order.
```

Don't mix Lisks solution with the general explanation of DPoS here. If you want you can add a separate sentence `*Delegated-Proof-of-Stake* is the consensus mechanism used on the Lisk Mainchain.` at the end of the DPoS explanation.

> The delegates are actually normal Lisk accounts that have performed a delegate registration transaction.

Reviewer: Again, don't talk about the Lisk solution here, just explain general DPoS
e.g.
```suggestion
Every user account can register as delegate by spending a certain amount of tokens.
```

The whole process can be considered as the most democratic method of voting, as it does not rest on the users that possess the highest amount of tokens, as just because a particular user has many tokens, that does not authorize this user to validate and confirm transactions.
All the users that maintain tokens in their accounts are able to select a group of delegates to perform this task.

> Furthermore, with this stake-weighted {url_voting-mechanism}[voting mechanism^] DPoS has the advantage of being able to execute transactions and verifications much faster than PoS or PoW, as there are a limited number of validators.

Reviewer: Do you have a proof for this? Please link/explain here. Also, this could be moved under the advantages section for DPoS

Furthermore, due to its decentralized nature, whereby a majority of nodes must achieve consensus before any new blocks can be added to the chain, this makes it almost impossible for any malicious actors to alter or manipulate the data added to the blockchain.

To summarise, a {url_p2p_networks}[P2P network architecture^] offers numerous benefits as compared to a traditional client-server-based network and results in greater freedom, improved security, and enhanced decentralization.

=== Unstructured P2P networks

In an unstructured P2P network there is no organization, hence the nodes are able to communicate and connect randomly.
Such networks are more suited to social platforms, or any type of system which experiences high churn rates.
However, such networks require a high amount of processing power and may incur long delays.

=== Structured P2P networks

A structured network can be considered the opposite of an unstructured network and is organized in a manner whereby the nodes are able to utilize efficient search mechanisms.
This can be performed by the nodes using a hash function.

> However, although they are considered to be more efficient, they can be considered somewhat centralized.

Reviewer: Why are you explaining these 2 different P2P network types here? What is the connection to blockchain? Please provide more context here for the reader.

Author: Yes, good point. I don't think we need to delve into the types of p2p networks here, so I have removed them 👍

== Cryptography

Cryptography is not a new concept and ultimately is used to ensure secure communication between 2 parties can be established over an unsecured connection.
This can be accomplished by using an encryption mechanism as depicted in the example diagram below, whereby a plain text document is sent as a ciphered version to the recipient.
The recipient can only decrypt the ciphered text if he or she holds the identical symmetrical encryption key.

> Therefore, this can be transmitted over an unsecured medium, as regardless of any third party or malicious actors intercepting this, they would not be able to decrypt the text, rendering it useless to them.

Reviewer: I wouldn't explain it here, as this can be explained in the section about asymmetric encryption. Can be removed

Author: Good point ..Done 👍

* Where and how is this used in blockchain?

Cryptography plays an important role in blockchain in maintaining security and is the underlying technology for securing wallets and performing transactions.

> For example, When creating a wallet on a blockchain, a public-secret key pair will be generated.

Reviewer: Here you talk only about asymmetric encryption. Where are the other encryptions used and why?

Author: Done 👍

However, it is important to note that there are two methods of cryptography algorithms that are deployed within the blockchain, which are the asymmetric model described here, and the hashing function described in the following paragraph below.

As discussed, the encryption techniques play an important role here, as it is critical for a blockchain to use the most secure methodology, which ensures the key functions can be securely managed, such as transaction authentication, digital signatures, and identity management.
It is quite common with regard to the management of cryptocurrencies that with the asymmetric model of encryption, the public key is generally the actual address that 'contains' the tokens or coins, and is publicly viewable.
Therefore, when a transaction is created, this has to be digitally signed with the secret key.
Once this has been received by the blockchain network, it can be verified with the same public key, therefore this proves the authorization of the transaction is genuine by the owner of the secret key, or any entity with knowledge of the secret key.

> Hence, transactions can be authenticated without having to reveal the identity of the account owner.

Reviewer: Is the identity of the transaction sender not always transparant/public?

Reviewer 2: good point

Author: Have reworded this to make it a bit clearer 👍

Subsequently, the public key is used for identity management and the account address, and the secret key is derived from the passphrase of the user account to access the funds in the account associated with the address, and hence is able to authorize and then perform any actions required.

image:intro/asymmetric-encryption.png[]

> === Hash Functions

Reviewer: What are the main benefits & characteristics of hashing?

Why is it important in blockchain context? Please explain and provide examples.

Author: Reworded above, and added in more content to cover hashing. 👍

Reviewer: Please provide a concrete example for hashing in blockchain context. The most prominent example is probably hashing the previous block and adding the hash to the block header of the next block, to ensure immutability

Author: OK, I have added a bit more content here and had already given examples (Bitcoin and Ethereum) 👍

This function does not utilize any keys, as it takes the contents of the plain text and deploys a cipher, which is used to generate a hash value of a fixed length from the plain text.
Hence, it is virtually impossible for the contents of this plain text to be unraveled from the cipher text.
Therefore, {url_hashing}[hashing^] outputs can be used to efficiently verify the inputs without revealing the input itself.
As described earlier on the previous page, it is also deterministic and is the key component in providing immutability of the blockchain data as well, which as we have learned is highly beneficial.

> Lisk maintains a cryptography package that contains all the cryptographic functionalities required when interacting with the Lisk ecosystem and can be used on both the server and client-side.

Reviewer: Not sure if we should mention it here, but if, add a link to the package reference

Author: I think it's good to mention this, and have added a link as requested 👍

Once a block of data is added to the blockchain it will remain publicly viewable to anyone at any time, assuming it is not a permissioned or private blockchain.
This is a highly beneficial system as in general as long as the blockchain is still active, and the history is still retrievable, then this results in ensuring that all the records are able to be verified.
For example, these could be virtually anything such as medical records, financial transactions, property rights, etc.

To start to break down how the blockchain works, and before we delve into the detailed contents of a block, firstly we need to understand how the blocks are linked together, this is accomplished by using cryptography.
Therefore, each block contains a cryptographic hash of the previous block, the transaction data,  and a timestamp.
This will be explained in more detail further in this section.
As previously mentioned once finalized, the blocks are immutable and cannot be changed retroactively.
The only way this could be achieved would be by altering the subsequential blocks, which would require the consensus of the entire network.

> == Public, private, and permissioned blockchains

Author: Done 👍

> == Public, private, and permissioned blockchains

Author: I have restructured it somewhat, and also built a table that covers the missing points.

In addition, permissioned blockchains are basically a mixture of private and public blockchains and may be customized to suit the requirements.
Examples of private blockchains are usually internal blockchains run by private businesses.
In addition, business-to-business virtual currency exchanges such as Hyperledger or Ripple are also private chains.
On the other hand, public blockchains are the likes of Ethereum and Bitcoin whereby it is possible for anyone to access the chain and add and transact blocks.
Finally, regarding permissioned blockchains, certain banks operate permissioned blockchains, which are run via a designated number of internal nodes that can be used to track money transactions.
Other examples are the likes of {url_quorum}[Quorum^] and {url_corda}[Corda^], as generally, they offer a more cost-effective solution coupled with a better governance structure.

In summary, although public and private blockchains have gained more traction and have become more widely known, the permissioned blockchain is now seeing increasing usage, as one of the key attributes is that it allows limited activities to be controlled, even by external entities.

> == Key characteristics of blockchain

Author: I have added an explanation for Validators, although it seems we have covered the explanations of other terminologies, and as you mentioned also on the previous 'What is blockchain' page.

There are 4 core characteristics attributed to the blockchain which are covered below:

* Decentralization – This can be defined as storing the blocks over a network of nodes, and not in any one central location.
The contents of each block in a public blockchain can be accessed by any node on the network, as they are essentially network copies that are present on all participating nodes.
Therefore, there is no central entity that holds any control to change each block in the chain, so this results in decentralization.

> * Consensus Driven – A consensus model independently verifies each block on the blockchain, and provides certain rules for validating a block, which may utilize key resources such as computing power, in order to demonstrate that this has been accomplished.

Author: Good point, I have removed this 👍

For example, with regard to Bitcoin, this is known as the mining process.
There are different types of consensus models which will be covered further in this section, however, suffice to say that a consensus model is able to function independently, and does not require any form of central authority.

* Immutability – A blockchain can be considered as an unchangeable and irreversible record of transactions.

> Hence, once a block is finalized and added to the blockchain it ensures any transaction in that block is completely trustable.

Author: Good point, added in a brief explanation 👍

* Transparency – A public blockchain by its very nature provides total open transparency.
A full transaction history can be viewed by any party, as it is in effect an open file.
Therefore, another highly beneficial advantage is that a blockchain creates provenance, as all transactions can be easily audited from the first to the last entry.

== Consensus mechanisms

It is important to understand consensus within the blockchain ecosystem.
Basically, a consensus mechanism is a methodology deployed by nodes to agree on, and validate the data content of a block before it becomes part of a blockchain.

> The block containing this data is then distributed to all the nodes within the network.

Author: Fair point...now done 👍

The consensus algorithm is responsible for ensuring that the exact same copy of the data encapsulated in each block exists within each node.
To put it in a nutshell, a consensus mechanism can be thought of as a necessary procedure that results in ensuring all peers of a blockchain network come to a common agreement on the current state of the network.

> Various consensus mechanisms exist today, however, the 4 main mechanisms that are widely used consist of the *Proof-of-Work*, *Proof-of-Stake*, *Delegated-Proof-of-Stake*, and *Byzantine-Fault-Tolerance* mechanisms, which are explained below:

Author: Done 👍

=== Proof-of-Work (PoW)

image::intro/pow.png[ align="center" ,600]

Proof of Work is probably the most well-known and established consensus mechanism, as it is utilized by Bitcoin.
Basically, to create a block it requires the nodes to solve an arbitrary mathematical problem or complex equations, which prevents the network from being hacked, and also provides a high level of security.
The term 'Proof of Work' stems from how the crypto miners effectively 'prove' that they have accomplished the necessary tasks in order to create a correctly formed block of transactions that will be added to the blockchain.
In essence, this can be thought of as a competition whereby the miners are attempting to outdo each other and solve the complex mathematical equations, which in turn then allows them to add the next block to the blockchain.
Hence, this results in the miners receiving a reward for this in the form of tokens or coins.
A further more in-depth explanation of PoW can be found {url_pow}[here^].

* What are the advantages of PoW?

> The major advantage regarding Proof-of-Work is that it is highly scalable, and generally very reliable.

Author: I have now updated all the consensus mechanism sections to contain bullet point lists of the advantages and disadvantages.

This type of attack can corrupt the network as with such a high amount of mining power, they can mine faster than all other miners.
In addition, they can also halt the confirmation and order of new transactions resulting in the network being interrupted.
A more in-depth description of the well-known 51% attack can be found {url_51}[here^]
Additional attacks that may occur consist of {url_sybil}[Sybil attacks^], and the {url_ddos}[DDoS^] (Distributed denial of Service) attacks.
Basically, a Sybil attack is whereby the attacker is able to fill the network with users that he or she can control, and perform nefarious actions.
In essence, this consists of having multiple network nodes that can act in unison to control the PoW mechanism.
A DDoS attack is not specific to blockchain, although it involves the attacker sending vast amounts of data to a node, therefore, rendering it unable to process these transactions, at which point the attacker would then be able to send new nodes under his control to the network resulting in a Sybil attack as described above.

=== Proof-of-Stake (PoS)

> image::intro/pos-v3.png[ align="center" ,600]

Author: This has been updated. The image works for me, however, I have reinserted it again.

> To explain PoS briefly, users are able to stake an asset/token which in turn opens up the possibility to be chosen as a validator of a new block, which then allows transaction fees, or newly minted tokens to be collected from the block as a reward.

Author: From my understanding, any tokens that are just held in a wallet, are not actually staked tokens..You have to actively stake them to reap any rewards.

Staking is defined as a number of tokens/assets that are held by the node.
The proof-of-stake-mechanism uses an algorithm designed to select users that have the highest stakes as validators, this results in motivating the highest stakeholders to ensure a transaction is executed, as logic dictates that the users with the highest amount of tokens or coins have the most to lose, therefore it is in their interest to ensure the network continues to grow.
This is highly beneficial for consensus building and eliminates the need for complex mathematical calculations, hence reducing the overall computing power and energy required.

> Alternatively, the nodes that hold the most tokens/assets have the power to validate the transactions, which in turn results in this node receiving a reward, usually in the form of newly minted tokens.

Author: I have reworded the part on 51% attacks. 
I also removed the sentence above 👍

Further more detailed information regarding PoS can be found {url_pos}[here^].

* What are the advantages of PoS?

Proof-of-Stake is far more efficient than PoW as it does not require any energy intensive computer hardware, and does not require such complex cryptographic mathematical problems to be solved to complete the mining process.
In addition, PoS is more energy efficient than PoW, which means new blocks can be added to the chain with minimal effort and energy required.

* What are the disadvantages of PoS?

One of the well known disadvantages relates to the fact that the mining power in PoS is determined by the number of tokens that a validator has staked, therefore it is often said that this tends to benefit the more wealthy participants.
Hence, users that stake more tokens have a higher chance of being chosen to forge new blocks.

> To explain PoS briefly, it allows the users to stake an asset/token which in turn opens up the possibility to be chosen as a validator of a new block, which then allows transaction fees, or newly minted tokens to be collected from the block as a reward.

Author: Yes, I noticed this and removed it earlier 👍

This is highly beneficial for consensus building and eliminates the need for complex mathematical calculations, hence reducing the overall computing power and energy required.
Alternatively, the nodes that hold the most tokens/assets have the power to validate the transactions, which in turn results in this node receiving a reward, usually in the form of newly minted tokens.
Further more detailed information regarding PoS can be found {url_pos}[here^].

=== Delegated-Proof-of-Stake (DPoS)

image::intro/dpos-v1.jpeg[ align="center" ,600]
// This image is temporary - awaiting updated DpOS image
DPoS works in a similar fashion to PoS, however, one of the key differences is that it utilizes a delegation and voting mechanism, which in turn incentivizes the users to use their staked collateral to secure the network.
To allow this mechanism to be both efficient and effective at performing transaction validations, various different components of delegated proof of stake exist.

> DPoS is generally considered to be the most advantageous mechanism, as it negates most of the limitations that exist with PoW and PoS as previously described.

Author: Well, we now list the highest amount of advantages when using the DPOS mechanism as compared to PoW & PoS 👍

> DPoS utilizes a unique election system that is able to select users that can perform block verification.

Author: Done 👍

> In this system, generally users, or as they are known in PoS and DPoS, *delegates* are voted in based on their reputation.

Author: Yes, it is now 👍

> Delegates can also be thought of as witnesses or users producing blocks.

Author: I have removed this line. In addition, I have added an explanation to explain what validators are, before any description of any process is described which pertains to validators. 👍

With DPoS, it is possible to vote on delegates by entering your tokens into a staking pool and linking them to a specific delegate.
In DPoS a limited number of delegates exist (usually from around 20 to 100), and the delegates are voted in by other users.
So the users that are chosen to forge each block, may not be the same users who are chosen to forge the preceding block.
In addition, it is this limited number of delegates that oversee the governance of a blockchain deploying the DPoS consensus mechanism.
Each user who holds a minimum of one token/coin with the DPoS blockchain has the ability to vote specifically for the delegates that they want to perform the transaction validations.
Dependent on these votes, certain delegates are allowed to add blocks to the blockchain in a specific order.

> Every user account can register as delegate by spending a certain amount of tokens.

Author: Am not clear on this one.. Do you mean to remove this entire paragraph?

Finally, this also results in higher scalability.

* What are the disadvantages of DPoS?

DPos is a partially centralized system, therefore the delegates with more tokens tend to have more power in the network.
DPoS can be classed as being less centralized than other consensus protocols such as PoS and PoW.
However, with DPoS effectively a limited number of users retain control of the network. One criticism that is often levied at DPoS is that it sacrifices decentralization for scalability.
Furthermore, it can be vulnerable to attacks as often there are only minimal participants in charge of keeping the network functional.

== Byzantine-Fault-Tolerance (BFT)

The BFT mechanism was designed in a manner whereby it is able to tolerate failures in the network, coupled with being able to withstand malicious attacks and corrupted data.
In a nutshell, the BFT mechanism ensures that the same guaranteed data is received by every node present in the network.

> In essence, it allows consensus to be reached regardless if some of the nodes fail.

Author: Maybe I am missing something, but I cannot see where this is duplicated?

> In essence, it allows consensus to be reached regardless if some of the nodes fail.

Reviewer: Imho the line `231` just before is stating something similar, isn't it? Also, both sentences try to summarize, indicated by `In a nutshell ...` and `in essence ...` so I would only keep one of them.

> In essence, it allows consensus to be reached regardless if some of the nodes fail.

Author: Ah yes...Good point Mona! I have condensed this into one sentence now 👍

Firstly, there are 3 key features whereby BFT is able to improve the blockchain, and they are listed below:

* Safety: If 2 conflicting blocks occur on the network, then assuming two-thirds of the active delegates adhere honestly to the protocol, these 2 conflicting blocks will not be finalized simultaneously on the blockchain.

* Accountability: In the case whereby the protocol is violated by the delegate, they will be held responsible for this.
The key requirements for BFT must be accomplished by the nodes within a blockchain network, therefore, it is imperative they are deterministic and must begin with the same state for practical BFT.

* Liveness: New blocks can still be finalized on the network, even in the case whereby one third of the active delegates are offline.

> The process of being able to tolerate a number of Byzantine process failures or crashes, whilst ensuring the network continues to function is defined as Byzantine fault tolerance.

Author: Done 👍

To achieve BFT consensus the following requirements must be met:

1. Termination: Every known faulty process must conclude with an end result or output.
2. Agreement: The same output is decided by every non-faulty process.
3. Validity: Every process starts with the same input.
4. Integrity: The consensus value and all non-faulty process decisions achieved in point 2 above, need to have been put forward by some non-faulty process.

> == Peer-2-Peer networks

Author: Yes, that is a fair point. Although it just highlights that P2P is a decentralized system and beneficial for blockchain usage..We can remove it if you wish, or alternatively, I could add it in to bolster the Decentralization paragraph on the 'What is blockchain' page...What do you think Mona..?

> == Peer-2-Peer networks

Reviewer: Yes, sounds good! Let's remove it here for now, and add the missing parts to "What is blockchain?" page 👍

> == Peer-2-Peer networks

Author: Excellent! My thoughts exactly...I have now removed it, and will then create an issue later just to add this in to the 'What is blockchain?' page 👍

The usage of a peer-to-peer model in blockchain offers various benefits as compared to a standard client-server-based system.
By its very nature, in a P2P decentralized system, assets cannot be frozen or manipulated by a centralized entity.
Furthermore, due to its decentralized nature, whereby a majority of nodes must achieve consensus before any new blocks can be added to the chain, this makes it almost impossible for any malicious actors to alter or manipulate the data added to the blockchain.

To summarise, a {url_p2p_networks}[P2P network architecture^] offers numerous benefits as compared to a traditional client-server-based network and results in greater freedom, improved security, and enhanced decentralization.

== Cryptography

Cryptography is not a new concept and ultimately is used to ensure secure communication between 2 parties can be established over an unsecured connection.

> This can be accomplished by using an encryption mechanism as depicted in the example diagram below, whereby a plain text document is sent as a ciphered version to the recipient.

Reviewer: What diagram? I think the whole sentence can be removed

Author: Yes, fair point. I have removed this sentence 👍

* Where and how is this used in blockchain?

Cryptography in blockchain consists of 3 different types, symmetric, asymmetric, and cryptographic hashing and are explained further in this section.
These play an important role in blockchain in maintaining security and is the underlying technology for securing wallets and performing transactions.
For example, When creating a wallet on a blockchain, a public-secret key pair will be generated.

* Why is it used and what are the advantages?

Cryptography is used simply to secure the various transactions occurring on the blockchain network and to verify the transactions such as minting or transferring tokens or coins.
Furthermore, being able to store and protect large amounts of transactions, and provide protection from hackers or malicious actors is considered highly advantageous.

To delve a bit further into cryptography in blockchain, it is helpful to be aware of the 3 types of cryptography deployed today.

> These can be broken down into the following three types:

Reviewer: These sections do not answer where which type of cryptography is used in particular. Please make sure to include this information, and please try to highlight on this information, so that it is not part of one big text block, which makes it hard to find.

Where exactly which cryptography is used and why this is used here, is the most interesting part about the whole topic, so this information should be transmitted to the reader as clearly as possible.

It is less important to explain the mechanism themselves in much detail, because there are a lot of great online resources that we can link to, if people are unaware of it.

Author: I have restructured this to include, 'How', 'Why', and 'Where' it is used. In addition, I have added a link.

With regard to a blockchain system, it can be deemed as a deterministic, replicated state machine.

image::intro/state-machine.png[]

The state transition refers to the changes that occur in the state machine after a specific event has occurred.
Therefore, it is best thought of as a reference to the process of moving from 1 state to another via transactions.
Although it should be noted that even a block that does not contain any transactions also has the possibility to alter the state of the blockchain.
Furthermore, it maintains a temporary state that exists during the processing of a block.

Firstly, the xref:{url_account-store}[account store] handles token transfers, keys, and registering delegates.
Secondly, the chain state store is responsible for the delegate vote weights, the block headers of the 3 previous rounds, the network identifier, the total fees burnt, and finally the rewards for the last block.
Thirdly, the consensus store contains the validator information and the finalized block height, including the BFT voting ledger, and furthermore, regarding the BFT, the consensus store holds the internal state.

> Finally, this is explained in more depth in the xref:{state_store}[Modules and Assets] page covering the state store changes and execution logic.

Author: Done 👍

Subsequently, the public key is used for identity management and the account address, and the secret key is derived from the passphrase of the user account to access the funds in the account associated with the address, and hence is able to authorize and then perform any actions required.
The recipient can only decrypt the ciphered text if he or she holds the identical symmetrical encryption key.
Therefore, this can be transmitted over an unsecured medium, as regardless of any third party or malicious actors intercepting this, they would not be able to decrypt the text, rendering it useless to them.

image:intro/asymmetric-encryption.png[]

=== Hash Functions
This function does not utilize any keys, as it takes the contents of the plain text and deploys a cipher, which is used to generate a hash value of a fixed length from the plain text.
Hence, it is virtually impossible for the contents of this plain text to be unraveled from the cipher text.
Therefore, {url_hashing}[hashing^] outputs can be used to efficiently verify the inputs without revealing the input itself.

> Hashing the previous block and then adding the hash to the block header of the next block ensures immutability, and in the context of blockchain, the hashing function properties are used within the consensus mechanism.
> Blockchains are also dependent on hashing, and the well-known SHA-256 (Secure Hash Algorithm), cryptographic hash function is used extensively in blockchain technology, including Bitcoin.

Reviewer: ```suggestion
In the context of blockchain, the hashing function properties are used mainly within the consensus and application domains.
A popular use case for hashing in blockchain is to create a hash of the previous block, and to include that hash in the next block.
This way, a block is cryptographically connected to all its former blocks, forming the classical chain of blocks.
Even the smallest change in one of the former blocks would create a very different hash of that particular block, which wouldn't match with the hash in the header of the next block.
This would break the chain of blocks immediately.
Therefore, hashing is an important factor to ensure the *immuability* of the blockchain.
```

A hash basically converts the required data into a string of characters.
Hashing is able to store data efficiently, as the hash is of a fixed size, and in addition, provides security through encryption.
Other hashing functions are also used in the blockchain world, for example, Ethereum deploys the Keccak-256 cryptographic hash function.
There are numerous benefits of hashing, as it meets the encrypted requirements demanded by a blockchain network.
Its characteristics consist of being able to accept a message of any length, whilst producing a fixed-length message digest.
Furthermore, the hash is irreversible, ensuring it is impossible to generate any message from the message digest.

> As described earlier on the previous page, it is also deterministic and is the key component in providing immutability of the blockchain data as well, which as we have learned is highly beneficial.

Reviewer: Please add a link to the previous page. Also, it is better to mention the page name directly, instead of `previous page`

Author: Done 👍

Author: Regarding the v5 terms, I have removed the old 'account store', 'chain state store', and 'consensus store' terms, and just described the general process. Let me know your thoughts if this works, otherwise we could always just remove this paragraph?

= How blockchain works
Chris Braithwaite <christopher.braithwaite@lightcurve.io> Mona Bärenfänger <mona@lightcurve.io>
:description: The How blockchain works page describes in more detail the functionalities of a blockchain.

> :toc: preamble

Reviewer: ```suggestion
:toc:
```

== State machine

A State machine is considered to be a concept whereby the definition relates to a machine that can have multiple states, however only one state is possible at any one given time.
Hence, a state, in this case, refers to the current state of the blockchain system and the transactions are a way to transition from one state to another
With regard to a blockchain system, it can be deemed as a deterministic, replicated state machine.

image::intro/state-machine.png[]

The state transition refers to the changes that occur in the state machine after a specific event has occurred.
Therefore, it is best thought of as a reference to the process of moving from 1 state to another via transactions.
Although it should be noted that even a block that does not contain any transactions also has the possibility to alter the state of the blockchain.
Furthermore, it maintains a temporary state that exists during the processing of a block.

Firstly, the token transfers, keys, and registering delegates are handled, followed by the delegate vote weights, the block headers of the 3 previous rounds, the network identifier, the total fees burnt, and the rewards for the last block.

> Finally, then the validator information, including the finalized block height, and the BFT voting ledger are then complete.

Reviewer: Please remove, as this is Lisk-specific

Author: Done 👍

These play an important role in blockchain in maintaining security and are the underlying technology for securing wallets and performing transactions.
For example, When creating a wallet on a blockchain, a public-secret key pair will be generated.

* Why is it used and what are the advantages?

Cryptography is used simply to secure the various transactions occurring on the blockchain network and to verify the transactions such as minting or transferring tokens or coins.
It is mainly deployed in the application and consensus layers of the blockchain. Furthermore, being able to store and protect large amounts of transactions, and provide protection from hackers or malicious actors is considered highly advantageous.

* Where is cryptography used in blockchain?

Most cryptocurrencies utilize asymmetric encryption. For example, Ethereum uses asymmetric cryptography with the public-private key pair as described further in this section.

To delve a bit further into cryptography in blockchain, more detailed information can be found {url_crypto}[here^].
However, it is helpful to be aware of the 3 types of cryptography deployed today.

> These can be broken down into the following three types described below:

Reviewer: This section seems to be duplicated content with lines 276-280, would remove.

Author: Done 👍

Symmetric cryptography is used in the banking sector, a good example being card transaction payment applications.
Examples of some of the most widely used symmetric encryption algorithms are AES, Blowfish, and RC6.
However, although it offers secure protection, it is necessary that all parties involved have to exchange the secret key that has been used to perform the encryption before it can be decrypted.
Nevertheless, blockchain uses even a more enhanced encryption methodology known as Asymmetric Encryption, as described in the following paragraph.

image:intro/symmetric-encryption.png[]

=== Asymmetric Key Encryption

This type of encryption functions by using a pair of keys.
This comprises an encryption key, and a decryption key, and is more commonly known as a public key and a private (or secret), key.
Basically, the algorithm deployed for this method generates both a secret, key and a unique public key.
The secret key as its name implies, is kept secret, and the public key is openly shared.
Furthermore, the asymmetric encryption method has an additional element of security, although the symmetric method of encryption is faster, nevertheless, they are both very effective.

> However, it is important to note that there are two different methods of cryptography algorithms that are deployed within the blockchain, and these are the asymmetric model and the hashing function.

Reviewer: Would remove this line

Author: Done 👍

Blockchains are also dependent on hashing, and the well-known SHA-256 (Secure Hash Algorithm), cryptographic hash function is used extensively in blockchain technology.
A hash basically converts the required data into a string of characters.
Hashing is able to store data efficiently, as the hash is of a fixed size, and in addition, provides security through encryption.
Other hashing functions are also used in the blockchain world, for example, Ethereum deploys the Keccak-256 cryptographic hash function.
There are numerous benefits of hashing, as it meets the encrypted requirements demanded by a blockchain network.
Its characteristics consist of being able to accept a message of any length, whilst producing a fixed-length message digest.

> Furthermore, the hash is irreversible, ensuring it is impossible to generate any message from the message digest.

Reviewer: this paragraph seems to belong to the explanation of hashing below

Author: Yes indeed you are correct Mona, and I have already moved it now 👍
